handler/src/main/java/io/netty/handler/ssl/ApplicationProtocolNegotiationHandler.java

```diff
@@ -0,0 +1,127 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+/**
+ * Configures a {@link ChannelPipeline} depending on the application-level protocol negotiation result of
+ * {@link SslHandler}.  For example, you could configure your HTTP pipeline depending on the result of ALPN:
+ * <pre>
+ * public class MyInitializer extends {@link ChannelInitializer}&lt;{@link Channel}&gt; {
+ *     private final {@link SslContext} sslCtx;
+ *
+ *     public MyInitializer({@link SslContext} sslCtx) {
+ *         this.sslCtx = sslCtx;
+ *     }
+ *
+ *     protected void initChannel({@link Channel} ch) {
+ *         {@link ChannelPipeline} p = ch.pipeline();
+ *         p.addLast(sslCtx.newHandler(...)); // Adds {@link SslHandler}
+ *         p.addLast(new MyNegotiationHandler());
+ *     }
+ * }
+ *
+ * public class MyNegotiationHandler extends {@link ApplicationProtocolNegotiationHandler} {
+ *     public MyNegotiationHandler() {
+ *         super("http/1.1");
+ *     }
+ *
+ *     protected void configurePipeline({@link ChannelHandlerContext} ctx, String protocol) {
+ *         if ("h2".equals(protocol) {
+ *             configureHttp2(ctx);
+ *         } else if ("http/1.1".equals(protocol)) {
+ *             configureHttp1(ctx);
+ *         } else {
+ *             throw new IllegalStateException("unknown protocol: " + protocol);
+ *         }
+ *     }
+ * }
+ * </pre>
+ */
+public abstract class ApplicationProtocolNegotiationHandler extends ChannelInboundHandlerAdapter {
+
+    private static final InternalLogger logger =
+            InternalLoggerFactory.getInstance(ApplicationProtocolNegotiationHandler.class);
+
+    private final String fallbackProtocol;
+    private SslHandler sslHandler;
+    private boolean configured;
+
+    /**
+     * Creates a new instance with the specified fallback protocol name.
+     *
+     * @param fallbackProtocol the name of the protocol to use when
+     *                         ALPN/NPN negotiation fails or the client does not support ALPN/NPN
+     */
+    protected ApplicationProtocolNegotiationHandler(String fallbackProtocol) {
+        this.fallbackProtocol = ObjectUtil.checkNotNull(fallbackProtocol, "fallbackProtocol");
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        // FIXME: There is no way to tell if the SSL handler is placed before the negotiation handler.
```

Comment:
Hmm.. maybe. I'm not really sure, so I'd like to wait until someone reports a real world case. :-)